### ISSUE TITLE
MokoResources Escape Backslashes in moko-resources-generated.js

### ISSUE THREAD

**Issue by wakaztahir** [icerockdev/moko-resources#409]
![Capture](https://user-images.githubusercontent.com/42442700/215316768-405c7ebe-f1ff-46b6-b1fd-0f09631ae618.PNG)

I fixed the last error by changing this line inside the `webpack.config.d/moko-resources-generated.js` of the both modules , here's the changed line

```
const mokoResourcePath = path.resolve("D:/Apps/My Apps/APP/web/build/generated/moko/jsMain/comwakaztahirsketchapp/res");
```

At the current time the line contains backslashes like this 

```
D:\Apps\My Apps\APP\web\build\generated\moko\jsMain\comwakaztahirsketchapp\res
```

The [third commit](https://github.com/jeran/moko-js-bug/commit/a40f03301e7d85975d6d84f236591783a85f7455) mentioned in https://github.com/icerockdev/moko-resources/issues/392 uses forward slashes in `moko-resources-generated.js`

There's also this issue https://github.com/icerockdev/moko-resources/issues/354

https://github.com/icerockdev/moko-resources/blob/df157ee999b787bf0d3628ab33747ce51b56286d/resources-generator/src/main/kotlin/dev/icerock/gradle/generator/js/JsMRGenerator.kt#L155

You should replace it with this

`const mokoResourcePath = path.resolve("${resourcesOutput.absolutePath.replace("\\","\\\\")}");`

**Comment by wakaztahir** [comment]
![Capture](https://user-images.githubusercontent.com/42442700/215315029-082959d5-4def-4606-8f6c-6ffca6dd773c.PNG)
I have found as you can see the reference and the file exists as well , I just don't know why this is not working

On Android I did have to add this code , for it to work

```
   sourceSets["main"].apply {
        res.srcDir(File(buildDir, "generated/moko/androidMain/res"))
    }
```

**Comment by wakaztahir** [comment]
After pasting the json to the specified location , as a workaround , Now I'm getting this error

![Capture](https://user-images.githubusercontent.com/42442700/215316768-405c7ebe-f1ff-46b6-b1fd-0f09631ae618.PNG)
Compiled with problems:X

ERROR in ./kotlin/Sketchable-sketchable-compose.js 363:16-80

Module not found: Error: Can't resolve 'localization/comwakaztahirsketchable_stringsJson.json' in 'D:\Programming\Android\Apps\My Apps\Sketchable\build\js\packages\Sketchable-web\kotlin'
Did you mean './localization/comwakaztahirsketchable_stringsJson.json'?
Requests that should resolve in the current directory need to start with './'.
Requests that start with a name are treated as module requests and resolve within module directories (node_modules).
If changing the source code is not an option there is also a resolve options called 'preferRelative' which tries to resolve these kind of requests in the current directory too.

**Comment by wakaztahir** [comment]
found this https://github.com/icerockdev/moko-resources/issues/392

**Comment by wakaztahir** [comment]
I have found that moko-resources doesn't work with multi-module projects

Problem #1 : The files don't get copied
Problem #2  : Relative paths problem (can't find a workaround for this one)
Problem #3 : You have to clean and rebuild multiple times or delete files manually for it to regenerate the files

This happens even after applying the plugin to the application module with a different package name , I thought it would fix the paths issue but it didn't

**Comment by wakaztahir** [comment]
Could you please provide a workaround for the paths problem

**Comment by wakaztahir** [comment]
I fixed the last error by changing this line inside the `webpack.config.d/moko-resources-generated.js` of the consuming module (main) , here's the changed line

```
const mokoResourcePath = path.resolve("D:/Apps/My Apps/APP/web/build/generated/moko/jsMain/comwakaztahirsketchapp/res");
```

At the current time the line contains backslashes like this 

```
D:\Apps\My Apps\APP\web\build\generated\moko\jsMain\comwakaztahirsketchapp\res
```

**Comment by wakaztahir** [comment]
The [third commit](https://github.com/jeran/moko-js-bug/commit/a40f03301e7d85975d6d84f236591783a85f7455) mentioned in https://github.com/icerockdev/moko-resources/issues/392 uses forward slashes in `moko-resources-generated.js`

**Comment by wakaztahir** [comment]
This does resolve the problem (changing backslashes to forward slashes in both module's moko-resources-generated.js) but the error shows up above the UI like this

![Capture](https://user-images.githubusercontent.com/42442700/215334373-b7d44c70-dd23-44d0-af43-c6a3dcf4ed22.PNG)

**Comment by wakaztahir** [comment]
##### Please fix the backslashes and change them to forward slashes

**Comment by wakaztahir** [comment]
`const mokoResourcePath = path.resolve("${resourcesOutput.absolutePath}");`

https://github.com/icerockdev/moko-resources/blob/df157ee999b787bf0d3628ab33747ce51b56286d/resources-generator/src/main/kotlin/dev/icerock/gradle/generator/js/JsMRGenerator.kt#L155

You should replace it with this

`const mokoResourcePath = path.resolve("${resourcesOutput.absolutePath.replace("\\","\\\\")}");`

**Comment by y9san9** [comment]
Thank you for the issue, I will investigate this in a few days

**Comment by wakaztahir** [comment]
Please fix this issue fast , This issue gives me a headache , every time I start the app , I have to escape the backslashes , I tried writing a gradle task , but it wasn't running after the generateMRjsMain in the subproject , plus gradle disables optimizations 

```
Execution optimizations have been disabled for task ':web:jsProcessResources' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: 'D:\Programming\Android\Apps\My Apps\Sketchable\web\build\generated\moko\commonMain\res'. Reason: Task ':web:jsProcessResources' uses this output of task ':web:generateMRcommonMain' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.5/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'D:\Programming\Android\Apps\My Apps\Sketchable\web\build\generated\moko\jsMain\comwakaztahirsketchapp\res'. Reason: Task ':web:jsProcessResources' uses this output of task ':web:generateMRjsMain' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.5/userguide/validation_problems.html#implicit_dependency for more details about this problem.
```

**Comment by wakaztahir** [comment]
```kotlin
// https://github.com/icerockdev/moko-resources/issues/409
tasks.register<Task>("modifyWebpackConfig"){
    dependsOn(":core-ui:generateMRjsMain")
    doFirst {
        val file = project.rootDir.resolve("core-ui/webpack.config.d/moko-resources-generated.js")
        if (file.exists()) {
            val text = file.readText()
            val startIndex = text.indexOf("const mokoResourcePath")
            val endIndex = text.indexOf(';', startIndex = startIndex)
            val toreplace = text.substring(startIndex, endIndex)
            if (!toreplace.contains("/") && toreplace.contains("\\")) {
                file.writeText(text.replaceRange(startIndex, endIndex, toreplace.replace("\\", "/")))
            }
        }
    }
}
tasks.getByPath(":core-ui:jsProcessResources").dependsOn(":core-ui:modifyWebpackConfig")
```

**Comment by wakaztahir** [comment]
@y9san9 Please fix this issue as soon as possible , Its a big hurdle in web dev of multiplatform projects that depend on moko resources

**Comment by wakaztahir** [comment]
Hi @y9san9 , When will this be fixed , I cannot add any resources or run web module on windows unless I escape backslashes by hand which is not a good solution , the task above fails since there's no way I can listen to file changes easily and update right after moko resources creates/updates the file , I'd be really thankful if you could fix this issue

**Comment by wakaztahir** [comment]
@y9san9 Hi , is this issue going to be fixed any soon ?

**Comment by wakaztahir** [comment]
I created a fork https://github.com/Qawaz/moko-resources which fixes this issue

The only thing you need to add is the github packages repository in the pluginManagement block and change the version in the classpath to 100.20.1 instead of 1.20.1 and you can use the fork

It works , I've tested it

**Comment by wakaztahir** [comment]
I would have to publish all the packages 😭

**Comment by wakaztahir** [comment]
managed to make a working fork , I posted the java publication instead of plugin publication , its working

**Comment by Alex009** [comment]
will be released in 0.21.0
